### DISCUSSION TITLE
Transfer Old Values

### DISCUSSION THREAD

**Discussion by maxnezdyur** [idaholab/moose#22825] [Q&A General] [answered]
Is there a way to transfer old values of a variable when using multiapp?

**Comment by hugary1995** [comment, marked as answer]
Yes, the trick is to use an AuxVariable to store the old value. Something like
```python
[Variables]
  [u]
  []
[]

[AuxVariables]
  [u_old]
    [AuxKernel]
      type = ParsedAux
      function = 'u'
      args = 'u'
      execute_on = TIMESTEP_END
    []
  []
[]
```

**Reply by maxnezdyur** [reply]
Ok, that's what I did before. I believe that fails when using more than one fixed point iteration. Since the parent app at the end of the first fixed point iteration will update u_old to be u, when it transfers again, it will send the same values as the current u. If only there was an option like FIXED_POINT_END. I am creating an OldAux AuxKernel that I will run at TimeStepBegin on the child app, but I feel like there might be a use for FIXED_POINT_END.

**Reply by hugary1995** [reply]
yeah, I don't like abusing TIMESTEP_END for FPI_END either. If someone has the time to fix it I'm all for it. But maybe @YaqiWang has some special needs for that?

**Reply by YaqiWang** [reply]
> the parent app at the end of the first fixed point iteration will update u_old to be u

Are you sure about this? You can use this to copy the old values into an aux var:

```
[AuxVariables]
  [u_old][]
[]

[AuxKernels]
  [copy_old]
    type = SelfAux
    variable = u_old
    v = u
    implicit = false
    execute_on = timestep_begin
  []
[]
```
I think it should work. Even though the aux kernel will be executed multiple times during fixed point iteration at every time step, it won't change `u_old`.

**Reply by YaqiWang** [reply]
For the real timestep_begin timestep_end execute flags, we basically derived a new executioner with two new custom execute flags and let object executed on those. `timestep_begin` and `timestep_end` is so widely used in MOOSE and apps, renaming them and checking all objects whether on timestep_begin or fixed_point_begin would be a real headache.

**Reply by maxnezdyur** [reply]
If I wanted my child app to run before my parent app, would your solution work? Does a parent app run all AuxKernels with `timestep_begin` before it transfers to the child app when using `TransientMultiApp` on `TIMESTEP_BEGIN`?

**Reply by GiudGiud** [reply]
if you wanted to do that you'd execute the auxkernel on timestep_end to make it work

**Reply by GiudGiud** [reply]
btw there is the SelfAux to do this shorter than a ParsedAux

**Reply by YaqiWang** [reply]
If the aux kernel and the multiapp are both on timestep_begin, multiapp/transfer will get executed first. They will not see the updated solution at the first fixed point iteration. The later iterations will see the updated solution and the solution will remain unchanged (I hope I understand this right). If you make the copy (aux kernel) on timestep_end with `implicit = true`, it will get executed on inside the current fixed point iteration, so it possibly won't work. If `implicit = false`, it will be the same as `implicit = true` plus on timestep_begin.

**Reply by maxnezdyur** [reply]
@GiudGiud  

Are you saying that this system would work? I think this would have a different problem. For the first FPI, it would send over "older" solution (for the new timestep), then for every other FPI, it would send over the correct old solution. 

```
[AuxVariables]
  [copy_old]
    type = SelfAux
    variable = u_old
    v = u
    implicit = false
    execute_on = timestep_end
  []
[]
```

**Reply by GiudGiud** [reply]
good point. 
Ok we do need a fixed_point_end, fixed_point_begin execute_on for this then

**Reply by GiudGiud** [reply]
https://github.com/idaholab/moose/pull/23065 is merged.

Please try it for this purpose if you have time

**Reply by maxnezdyur** [reply]
This works perfectly. Thanks!

**Reply by maxnezdyur** [reply]
I may have spoken too soon. I have a 3 tier multtiapp: solid->fluid->ale. I have a transfer between fluid and ale that only needs to happen at `MULTIAPP_FIXED_POINT_END` which should be the last fixed point before we move to the next timestep(that's what I thought it was). Below shows that transfers on `MULTIAPP_FIXED_POINT_END` are being run, but we continue with more iterations. 

```
fluid_domain0:   Computing Jacobian                                                                     [  6.11 s] [  -23 MB]
fluid_domain0:  3 Nonlinear |R| = 1.045989e-11
fluid_domain0: Finished Solving                                                                         [ 57.83 s] [-1404 MB]
fluid_domain0:  Solve Converged!
fluid_domain0: Finished Computing User Objects                                                          [  2.80 s] [  287 MB]
fluid_domain0: 
fluid_domain0: Transfers on MULTIAPP_FIXED_POINT_END To MultiApps:
fluid_domain0: ---------------------------------------------------------------------------
fluid_domain0: |      Name      |                  Type                   |  From  | To  |
fluid_domain0: ---------------------------------------------------------------------------
fluid_domain0: | push_indicator | MultiAppShapeEvaluationTransfer         | Parent | ALE |
fluid_domain0: | push_nodeid    | MultiAppGeneralFieldNearestNodeTransfer | Parent | ALE |
fluid_domain0: | push_vel_x     | MultiAppShapeEvaluationTransfer         | Parent | ALE |
fluid_domain0: | push_vel_y     | MultiAppShapeEvaluationTransfer         | Parent | ALE |
fluid_domain0: | push_p         | MultiAppShapeEvaluationTransfer         | Parent | ALE |
fluid_domain0: ---------------------------------------------------------------------------
fluid_domain0: 
fluid_domain0:   Transferring variables via finite element interpolation.                               [ 16.52 s] [  110 MB]
fluid_domain0:   Transferring variables via finite element interpolation.                               [ 16.54 s] [    9 MB]
fluid_domain0: ..                              [ 16.61 s] [    4 MB]
fluid_domain0: ..                              [ 16.65 s] [   31 MB]
fluid_domain0: Transfers on MULTIAPP_FIXED_POINT_END Are Finished
fluid_domain0: 
fluid_domain0: Finished Executing Transfers                                                             [ 67.36 s] [  286 MB]
    Finished Executing MultiApps                                                         [181.03 s] [ -286 MB]
  Finished Executing MultiApps                                                           [181.03 s] [ -285 MB]
Fixed point residual norm after TIMESTEP_END MultiApps: 7.550675e-01

 0 Picard |R| = 1.763312e-13
 1 Picard |R| = 5.563353e+01
 2 Picard |R| = 5.352814e+00
 3 Picard |R| = 7.550675e-01

Beginning fixed point iteration 3


Main app solve:

    Rebuilding Geometric Search Patches
      Finished Reinitializing Because of Geometric Search Objects                        [  2.89 s] [  141 MB]
    Finished Rebuilding Geometric Search Patches                                         [  2.89 s] [  141 MB]
 0 Nonlinear |R| = 7.550675e-01
  Linear solve converged due to CONVERGED_RTOL iterations 1
 1 Nonlinear |R| = 5.851079e-02
  Linear solve converged due to CONVERGED_RTOL iterations 2
 2 Nonlinear |R| = 5.160122e-03
```

**Reply by GiudGiud** [reply]
so it's between level 2 and 3?
This level is nested within the fixed point iteration between 1 & 2. So we're not going to be able to do that

You could use siblings transfer and have both fluid and ale be subapps of solid. Then you can choose to perform this transfer only once

**Reply by maxnezdyur** [reply]
Yes, it's between levels 2 and 3. Ahh, that makes a lot more sense. I'll have to look at the sibling sub-apps, thanks.
